lib/cretonne/src/loop_analysis.rs

```diff
@@ -67,14 +87,32 @@ impl LoopAnalysis {
         self.loops[lp].parent.expand()
     }
 
-    /// Determine if an Ebb belongs to a loop by running a finger along the loop tree.
+    /// Determine if an `Ebb` belongs to a loop by running a finger along the loop tree.
     ///
     /// Returns `true` if `ebb` is in loop `lp`.
     pub fn is_in_loop(&self, ebb: Ebb, lp: Loop) -> bool {
-        let ebb_loop = self.ebb_loop_map[ebb];
-        match ebb_loop.expand() {
-            None => false,
-            Some(ebb_loop) => self.is_child_loop(ebb_loop, lp),
+        match self.ebb_loop_map[ebb] {
+            EbbLoopData::NoLoop() => false,
+            EbbLoopData::Loop { loop_id, .. } => self.is_child_loop(loop_id, lp),
+        }
+    }
+
+    /// Determine which region of an `Ebb` belongs to a particular loop.
+    ///
+    /// If `ebb` belongs to `lp`, it returns `None` if the whole `ebb` belongs to `lp` or
+    /// `Some(inst)` where `inst` is the last instruction of `ebb` to belong to `lp`.
+    pub fn last_loop_instruction(&self, ebb: Ebb, lp: Loop) -> Result<Option<Inst>, ()> {
```

Comment:
If all of an EBB's instructions belong to the loop, it seems that this function will return None or the EBB terminator somewhat randomly.

Wouldn't it be more consistent if this never returned the terminator? Only return some instruction when the EBB is actually split between two loops?